Microsoft.Toolkit.HighPerformance/Memory/Memory2D{T}.cs

```diff
@@ -0,0 +1,238 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Buffers;
+using System.ComponentModel;
+using System.Diagnostics.Contracts;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using Microsoft.Toolkit.HighPerformance.Extensions;
+
+namespace Microsoft.Toolkit.HighPerformance.Memory
+{
+    /// <summary>
+    /// <see cref="Memory2D{T}"/> represents a 2D region of arbitrary memory. It is to <see cref="Span2D{T}"/>
+    /// what <see cref="Memory{T}"/> is to <see cref="Span{T}"/>. For further details on how the internal layout
+    /// is structured, see the docs for <see cref="Span2D{T}"/>. The <see cref="Memory2D{T}"/> type can wrap arrays
+    /// of any rank, provided that a valid series of parameters for the target memory area(s) are provided.
+    /// </summary>
+    /// <typeparam name="T">The type of items in the current <see cref="Span2D{T}"/> instance.</typeparam>
+    public readonly struct Memory2D<T> : IEquatable<Memory2D<T>>
+    {
+        /// <summary>
+        /// The target <see cref="Array"/> instance, if present.
+        /// </summary>
+        private readonly Array? array;
+
+        /// <summary>
+        /// The initial offset within <see cref="array"/>.
+        /// </summary>
+        private readonly IntPtr offset;
```

Comment:
should either be `int` or `nint`. `int` is probably better given it results in a smaller type. `IntPtr` is very unfriendly to work with particulary in indexing code

Reply:
We can't really use `int`, as it'd cause overflows on x64 and with large wrapped objects. And we can't use `nint` either as we don't have C# 9 yet at least in this PR. I have a separate PR for .NET 5 and C# 9 support, but we need to keep those separate. We might as well want to get back to this in the future and change this for `nint` or `nuint`, but for now `IntPtr` I'd say is the best choice. Also I don't really see the issue with it, we also have a dedicated `Unsafe.Add` overload that takes an `IntPtr` directly, so I'd say it's actually pretty convenient to use in this case 😊